conan/tools/qbs/qbsprofile.py

```diff
@@ -0,0 +1,315 @@
+            'profile': self._profile,
+            'default_profile': self._default_profile,
+        }
+        result = t.render(**context)
+        return result
+
+    def generate(self):
+        """
+        This method will save the generated files to the conanfile.generators_folder.
+
+        Generates the "qbs_settings.txt" file. This file contains Qbs settings such as toolchain
+        properties and can be imported using ``qbs config --import``.
+        """
+        check_duplicated_generator(self, self._conanfile)
+        self._check_for_compiler()
+        save(self.filename, self.render())
+
+    def _check_for_compiler(self):
+        compiler = self._conanfile.settings.get_safe('compiler')
+        if not compiler:
+            raise ConanException('Qbs: need compiler to be set in settings')
+
+        if compiler not in ['msvc', 'gcc', 'clang', 'apple-clang']:
+            raise ConanException(f'Qbs: compiler {compiler} not supported')
+
+    def _get_qbs_toolchain(self):
+        compiler = self._conanfile.settings.get_safe('compiler')
+        the_os = self._conanfile.settings.get_safe('os')
+        if the_os == 'Windows':
+            if compiler == 'msvc':
+                if msvs_toolset(self._conanfile) == 'ClangCL':
+                    return 'clang-cl'
+                return 'msvc'
+            if compiler == 'gcc':
+                return 'mingw'
+            if compiler == 'clang':
+                return 'clang-cl'
+            raise ConanException('unknown windows compiler')
+        if compiler == 'apple-clang':
+            return 'xcode'
+        # todo: other compilers?
+        return compiler
+
+    def _default_compiler_names(self, toolchain):
+        if toolchain == 'msvc':
+            return 'cl', 'cl'
+        if toolchain == 'clang-cl':
+            return 'clang-cl', 'clang-cl'
+        if toolchain in ('gcc', 'mingw'):
+            return 'gcc', 'g++'
+        if toolchain in ('clang', 'xcode'):
+            return 'clang', 'clang++'
+
+        # what about other toolchains? IAR, Cosmic have a bunch of compilers based on arch
+        return toolchain, toolchain
+
+    def _find_exe(self, exe):
+        if platform.system() == 'Windows':
+            exe = exe + '.exe'
+        if os.path.isabs(exe):
+            return exe
+        paths = self._build_env.get("PATH", "")
+        for p in paths.split(os.pathsep):
+            path = os.path.join(p, exe)
+            if os.path.exists(path):
+                return path
+        return shutil.which(exe)
+
+    def _toolchain_properties(self):
+        toolchain = self._get_qbs_toolchain()
+        the_os = self._conanfile.settings.get_safe('os')
+        vcvars_path = None
+        compiler = None
+        if the_os == 'Windows' and toolchain == 'msvc':
+            compiler = _find_msvc(self._conanfile)
+        elif the_os == 'Windows' and toolchain == 'clang-cl':
+            compiler, vcvars_path = _find_clangcl(self._conanfile)
```

Comment:
Is the standard practice in modern Qbs to have to define that full path to MSVC compiler in toolchain file for Qbs to find and use it? Or is there some default that Qbs will look for it automatically (kind of CMake), and then other cases require explicit definition (like when CMake with Ninja that can require vcvars to be activated to work)

Reply:
Both ways work. 

We have a `qbs setup-toolchains` tool that tries to detect **all** toolchains in a system and create a profile for each one. Than user can choose a profile which to use for their project, e.g. `qbs profile:myprofile`.

If no profile is specified (or, specifically, cpp.toolchainInstallPath is not set), Qbs will try to find compiler from PATH/predefined locations.

I suppose we can omit cpp.toolchainInstallPath and tell user to run that env script that is generated by Conan, that should also work (I din't test it though).

However, since we already need to pass a bunch of properties such as defines and stuff, it seemed like a sane idea to put everything, inc toolchain path in one place, in a profile.

Reply:
Sounds good, thanks for the explanation.